Switch chainstate db and cache to per-txout model

sipa opened the pull request:
Since Bitcoin Core v0.8, we've used a per-tx model for the chainstate database and its cache. This means that the database is effectively a map from txids to the *list* of unspent outputs for that transaction. This PR changes that to a map from outpoints ((txid,index) pairs) to just the *individual* unspent output for that outpoint.

The original reason for aggregating the outputs per transaction was to save space: this way, we could avoid duplicating the txid and transaction meta across the multiple outputs. However, LevelDB internally uses an encoding that omits repeated prefix bytes in keys, and because of that, duplicating the txids is not very significant.

There are many advantages for using a per-txout model:
* Simpler code.
* Avoiding the CPU overhead of deserializing and serializing the unused outputs.
* More predictable memory usage.
* More easily adaptable to various cache flushing strategies.

Downsides:
* Slightly larger on-disk representation, and sometimes larger in-memory representation (when there are multiple outputs for the same txid in the cache, which becomes optional).

This PR includes:
* The main commit from #10248. We'll be dropping tx nVersion from the undo data later, so reserializing data from disk won't roundtrip anymore. This change allows us to compute the checksum based on the written data rather than the reserialized data.
* #10249, as we'll be using some c++11 methods of std::unordered_map which aren't available in all versions of boost::unordered_map.
* #10250, as the new tests trigger the problem fixed there.
* A forward-compatible undo data format change (new versions can use old undo data, not the other way around).
* Switch to a new per-txout coinsview model with related database changes.
* Upgrade code to convert the old chainstate to a new chainstate (which is interruptible).
* Adapted unit and functional tests (thanks to @ryanofsky for adapting the unit tests).

Tests:
* All existing unit and functional tests pass.
* Manually verified that continuing a reindex started with old code succeeds (even when recovering from a crash in the middle of flushing while upgrading).

Commit links:

Preparation:

* e66dbde6d14cb5f253b3bf8850a98f4fda2d9f49: Add SizeEstimate to CDBBatch
* f54580e7e4f225bb615204daef32f72ab8688418: error() in disconnect for disk corruption, not inconsistency
* e484652fc36ef7135cf08ad380ea7142b6cbadc0: Introduce CHashVerifier to hash read data
* 7e0032290669fae5f52c256856c53038511c7db4: Add specialization of SipHash for 256 + 32 bit data
* d342424301013ec47dc146a4beb49d5c9319d80a: Remove/ignore tx version in utxo and undo
* c3aa0c11947dfd82702df276d39bb7f748dd83a1: Report on-disk size in gettxoutsetinfo
* 7d991b55dbf0b0f6e21c0680ee3ebd09df09012f: Store/allow tx metadata in all undo records

Switch to COutPoint/Coin instead of uint256/CCoins:
* 422634e2f5ac1ff74cd358144cecbac63007adc4: Introduce Coin, a single unspent output
* cb2c7fdac2dc74368ed24ae4717ed72178956b92: Replace CTxInUndo with Coin
* bd83111a0fcfdb97204a0180bcf861d3b53bb6c2: Optimization: Coin&& to ApplyTxInUndo
* 000391132608343c66488d62625c714814959bc9: Introduce new per-txout CCoinsViewCache functions
* f68cdfe92b37f5a75be612b7de3c1a03245696d0: Switch from per-tx to per-txout CCoinsViewCache methods in some places
* c87b957a32e03c09d410abadf661f87eb813bcdb: Only pass things committed to by tx's witness hash to CScriptCheck
* 8b3868c1b4bf89c41b26ecb3a4b7c3a2557e3868: Switch CScriptCheck to use Coin instead of CCoins
* 961e4839793f8b4ad37d29672faf1695ff6ec03a: Switch tests from ModifyCoins to AddCoin/SpendCoin
* 05293f3cb75ad08ca23cba8e795e27d4d5e4d690: Remove ModifyCoins/ModifyNewCoins
* 13870b56fcd0bfacedce3ae42a3de3d5e9dc7bc1: Replace CCoins-based CTxMemPool::pruneSpent with isSpent
* 4ec0d9e794e3f338e1ebb8b644ae890d2c2da2ee: Refactor GetUTXOStats in preparation for per-COutPoint iteration
* 50830796889ecaa458871f1db878c255dd2554cb: Switch CCoinsView and chainstate db from per-txid to per-txout

Cleanup:
* ce23efaa5c2e8e50744a896424b01052db34a3d6: Extend coins_tests
* 97072d6685564dd50aab4c145b1758ccc10863b3: Remove unused CCoins methods
* 41aa5b79a3d79f8afe4c556b4f14fb1dc0cc3f9f: Pack Coin more tightly
* b2af357f39c7d17ab6ddb2938531155bf90126ec: Reduce reserved memory space for flushing
* 8b25d2c0ce64d7f8577a0c2e601e505c9f1140bf: Upgrade from per-tx database to per-txout
* 580b023092a28f444068b53792eb542f9d5e6892: [MOVEONLY] Move old CCoins class to txdb.cpp
* 119e552f7ccd49c0137a3c6b4f94018a84d69620: Merge CCoinsViewCache's GetOutputFor and AccessCoin
* 73de2c1ff345ac38c098d7b1bef03176f3ea1f16: Rename CCoinsCacheEntry::coins to coin
* a5e02bc7f8a1af27fcafd892d8da651e8f1ab156: Increase travis unit test timeout
* 589827975f9f241e2f23eb674a7383592bff1cad: scripted-diff: various renames for per-utxo consistency

gmaxwell commented:
This is awesome.

sipa commented:
Rebased on top of new #10148.

sipa commented:
Rebased without #10148 at popular request.

paveljanik commented:
Few overrides needed probably:
```
+./coins.h:200:10: warning: 'BatchWrite' overrides a member function but is not marked 'override' [-Winconsistent-missing-override]
+    bool BatchWrite(CCoinsMap &mapCoins, const uint256 &hashBlock);
+         ^
+./coins.h:178:18: note: overridden virtual function is here
+    virtual bool BatchWrite(CCoinsMap &mapCoins, const uint256 &hashBlock);
+                 ^
+./coins.h:201:23: warning: 'Cursor' overrides a member function but is not marked 'override' [-Winconsistent-missing-override]
+    CCoinsViewCursor *Cursor() const;
+                      ^
+./coins.h:181:31: note: overridden virtual function is here
+    virtual CCoinsViewCursor *Cursor() const;
+                              ^
+2 warnings generated.
```

TheBlueMatt commented on test/functional/blockchain.py:
```diff
@@ -52,7 +52,6 @@ def _test_gettxoutsetinfo(self):
         assert_equal(res['txouts'], 200)
         assert_equal(res['bytes_serialized'], 13924),
         assert_equal(len(res['bestblock']), 64)
-        assert_equal(len(res['hash_serialized']), 64)
```
Why not just switch to hash_serialized_2?

TheBlueMatt commented on src/txdb.cpp:
```diff
@@ -242,3 +243,120 @@ bool CBlockTreeDB::LoadBlockIndexGuts(boost::function<CBlockIndex*(const uint256
+        boost::this_thread::interruption_point();
+        COldCoins oldcoins;
+        std::pair<unsigned char, uint256> key;
+        if (pcursor->GetValue(oldcoins) && pcursor->GetKey(key) && key.first == DB_COINS_OLD) {
```
Shouldnt the first two be an error and not a "done upgrading" result?

TheBlueMatt reviewed (commented):
Concept ACK. Briefly looked over most of the commits except the big one, which I only briefly skimmed, but am left wondering if there's an easier way to get that +650-881 commit in. At least the first 6 commits could be merged separately (I know the first two are already in another PR), but I'm not sure thats sufficient. Anyway, will circle back around to doing a full review as other things get merged.

sipa commented:
I'm splitting the commits up more. I've pushed one update already, but I'm working splitting the big commit further.

sipa commented:
Done. There is still one commit that does many things at once ("Switch CCoinsView and chainstate db to per-txout"), but splitting it is nontrivial. If requested, I can try splitting the database format change into a second commit, but that would require adding a bunch of conversion logic in the first commit that just gets removed in the second one. Thoughts?

paveljanik commented on src/undo.h:
```diff
@@ -12,53 +12,72 @@
+    template <typename Stream>
+    void Unserialize(Stream& s) {
+        // TODO: avoid reimplementing vector deserializer
+        uint64_t count;
```
init to 0 here please.

sipa committed:
Add SizeEstimate to CDBBatch

This allows estimating the in-memory size of a LevelDB batch.

sipa committed:
error() in disconnect for disk corruption, not inconsistency

The error() function unconditionally reports an error. It should only
be used for actually exception situations, and not for the type of
inconsistencies that ApplyTxInUndo/DisconnectBlock can graciously deal
with.

This also makes a subtle semantics change: in ApplyTxInUndo, when a
record with metadata is encountered (indicating it is the last spend
from a tx), don't wipe the CCoins record if it wasn't empty at that
point. This makes sure that UTXO operations never affect any other
UTXOs (including those from the same tx).

sipa committed:
Introduce CHashVerifier to hash read data

This is necessary later, when we drop the nVersion field from the undo
data. At that point deserializing and reserializing the data won't
roundtrip anymore, and thus that approach can't be used to verify
checksums anymore.

With this CHashVerifier approach, we can deserialize while hashing the
exact serialized form that was used. This is both more efficient and
more correct in that case.

sipa committed:
Add specialization of SipHash for 256 + 32 bit data

We'll need a version of SipHash for tuples of 256 bits and 32 bits
data, when CCoinsViewCache switches from using txids to COutPoints as
keys.

sipa committed:
Remove/ignore tx version in utxo and undo

This makes the following changes:
* In undo data and the chainstate database, the transaction nVersion
  field is removed from the data structures, always written as 0, and
  ignored when reading.
* The definition of hash_serialized in gettxoutsetinfo is changed to no
  longer incude the nVersion field. It is renamed to hash_serialized_2
  to avoid confusion. The new definition also includes transaction
  height and coinbase information, as this information was missing
  before.

This depends on having a CHashVerifier-based undo data checksum
verifier.

Apart from changing the definition of serialized_hash, downgrading
after using this patch is supported, as no release ever used the value
of nVersion field in UTXO entries.

sipa committed:
Store/allow tx metadata in all undo records

Previously, transaction metadata (height, coinbase or not, and before
the previous commit also nVersion) was only stored for undo records
that correspond to the last output of a transaction being spent.

This only saves 2 bytes per undo record. Change this to storing this
information for every undo record, and stop complaining for having it
in non-last output spends. This means that undo dat written with
this patch won't be readable by older versions anymore.

sipa committed:
Introduce Coin, a single unspent output

sipa committed:
Replace CTxInUndo with Coin

The earlier CTxInUndo class now holds the same information as the Coin
class. Instead of duplicating functionality, replace CTxInUndo with a
serialization adapter for Coin.

sipa committed:
Optimization: Coin&& to ApplyTxInUndo

This avoids a prevector copy in ApplyTxInUndo.

sipa committed:
Introduce new per-txout CCoinsViewCache functions

The new functions are:
* CCoinsViewCache::AddCoin: Add a single COutPoint/Coin pair.
* CCoinsViewCache::SpendCoin: Remove a single COutPoint.
* AddCoins: utility function that invokes CCoinsViewCache::AddCoin for
  each output in a CTransaction.
* AccessByTxid: utility function that searches for any output with
  a given txid.
* CCoinsViewCache::AccessCoin: retrieve the Coin for a COutPoint.
* CCoinsViewCache::HaveCoins: check whether a non-empty Coin exists
  for a given COutPoint.

The AddCoin and SpendCoin methods will eventually replace ModifyCoins
and ModifyNewCoins, AddCoins will replace CCoins::FromTx, and the new
AccessCoins and HaveCoins functions will replace their per-txid
counterparts.

Note that AccessCoin for now returns a copy of the Coin object. In a
later commit it will be change to returning a const reference (which
keeps working in all call sites).

sipa committed:
Switch from per-tx to per-txout CCoinsViewCache methods in some places

sipa committed:
Switch CScriptCheck to use Coin instead of CCoins

sipa committed:
Switch tests from ModifyCoins to AddCoin/SpendCoin

sipa committed:
Remove ModifyCoins/ModifyNewCoins

sipa committed:
Replace CCoins-based CTxMemPool::pruneSpent with isSpent

sipa committed:
Switch CCoinsView and chainstate db from per-txid to per-txout

This patch makes several related changes:
* Changes the CCoinsView virtual methods (GetCoins, HaveCoins, ...)
  to be COutPoint/Coin-based rather than txid/CCoins-based.
* Changes the chainstate db to a new incompatible format that is also
  COutPoint/Coin based.
* Implements reconstruction code for hash_serialized_2.
* Adapts the coins_tests unit tests (thanks to Russell Yanofsky).

A side effect of the new CCoinsView model is that we can no longer
use the (unreliable) test for transaction outputs in the UTXO set
to determine whether we already have a particular transaction.

sipa committed:
Remove unused CCoins methods

sipa committed:
Pack Coin more tightly

sipa committed:
Reduce reserved memory space for flushing

As the maximum amount of data that can be pulled into the cache due to
a block validation is much lower now (at most one CCoin entry per input
and per output), reduce the conservative estimate used to determine
flushing time.

sipa committed:
Upgrade from per-tx database to per-txout

sipa committed:
[MOVEONLY] Move old CCoins class to txdb.cpp

It's only used for upgrading from the old database anymore.

sipa commented:
Rebased and reset the author timestamps (so that GitHub shows them in logical order, $@#!*).

sipa committed:
Merge CCoinsViewCache's GetOutputFor and AccessCoin

They're doing the same thing now.

sipa committed:
Refactor GetUTXOStats in preparation for per-COutPoint iteration

TheBlueMatt committed:
Only pass things committed to by tx's witness hash to CScriptCheck

This clarifies a bit more the ways in which the new script execution
cache could break consensus in the future if additional data from
the CCoins object were to be used as a part of script execution.

After this change, any such consensus breaks should be very visible
to reviewers, hopefully ensuring no such changes can be made.

sipa commented:
Rebased.

ryanofsky commented on src/txmempool.cpp:
```diff
@@ -889,19 +888,23 @@ bool CTxMemPool::HasNoInputsOf(const CTransaction &tx) const
 CCoinsViewMemPool::CCoinsViewMemPool(CCoinsView* baseIn, const CTxMemPool& mempoolIn) : CCoinsViewBacked(baseIn), mempool(mempoolIn) { }
 
-bool CCoinsViewMemPool::GetCoins(const uint256 &txid, CCoins &coins) const {
+bool CCoinsViewMemPool::GetCoins(const COutPoint &txid, CCoin &coins) const {
```
In commit "Switch CCoinsView and chainstate db from per-txid to per-txout"

Maybe rename txid to outpoint in various places.

ryanofsky reviewed (commented):
utACK 493f03f02ab71376374bb851fe82241e5ee0a15f. This is a really nice change, and seems complete and ready to go as far as I could tell. I left a lot of comments and possible suggestions in my review, but they're all very minor, so feel free ignore anything not of interest.

sipa commented:
Updated and addressed many of @ryanofsky's review comments. I'll respond only individually to the ones I'm not addressing.

sipa commented on src/validation.cpp:
```diff
@@ -1278,11 +1278,9 @@ void UpdateCoins(const CTransaction& tx, CCoinsViewCache& inputs, CTxUndo &txund
-            }
+            CTxInUndo& undo = txundo.vprevout.back();
+            undo.nHeight = coins->nHeight;
+            undo.fCoinBase = coins->fCoinBase;
```
Older code requires that there are 0s for everything but the last spend of an output, in the disconnect code. That's what the "allow" refers to in the commit title: it allows nonzero metadata in any undo record.

ryanofsky replied:
In commit "Store/allow tx metadata in all undo records"

> Older code requires that there are 0s for everything but the last spend of an output, in the disconnect code. That's what the "allow" refers to in the commit title: it allows nonzero metadata in any undo record.

Could you clarify in the commit message what "won't be readable by older versions anymore" means? Will older software just log errors, or require re-indexing, or will it not work at all?

I understand that older code has a check for this condition here:
 https://github.com/bitcoin/bitcoin/blob/431a548faaf51c7a5fc89b6e479187a1c0e29805/src/validation.cpp#L1508. But I also thought this was one of the cases that DisconnectBlock would "graciously deal with" (according to earlier commit message).

sipa replied:
> Will older software just log errors and keep working, or require re-indexing, or will it not work at all?

Older code will fail during the start-up consistency check (the rollback test will notice entries with undo data present where it isn't expected), and fail to start.

> But I also thought this was one of the cases that DisconnectBlock would "graciously deal with"

Well, it does deal with it correctly (it's DISCONNECT_UNCLEAN case, not a DISCONNECT_FAILED one), but the code calling DisconnectBlock requires a DISCONNECT_OK.

In #10148, DisconnectBlock is called to rewind blocks for replay at startup, and there DISCONNECT_UNCLEAN is acceptable.

sipa commented on src/validation.cpp:
```diff
@@ -1016,11 +1016,14 @@ bool GetTransaction(const uint256 &hash, CTransactionRef &txOut, const Consensus
         int nHeight = -1;
-        {
+        size_t out = 0;
+        while (nHeight == -1 && out < 1024) { // Only try first 1024 outputs
```
This is very hard, unfortunately.

sipa commented on src/coins.h:
```diff
@@ -20,6 +20,81 @@
+    CCoin(CTxOut&& outIn, int nHeightIn, bool fCoinBaseIn) : fCoinBase(fCoinBaseIn), out(std::move(outIn)), nHeight(nHeightIn) {}
+    CCoin(const CTxOut& outIn, int nHeightIn, bool fCoinBaseIn) : fCoinBase(fCoinBaseIn), out(outIn), nHeight(nHeightIn) {}
+
+    void Clear() {
```
Not going to rename things all over the place now. Maybe I can add a commit at the end that fixes it up later.

sipa commented on src/coins.h:
```diff
@@ -312,28 +312,28 @@ class CCoins
 struct CCoinsCacheEntry
 {
-    CCoins coins; // The actual cached data.
+    CCoin coins; // The actual cached data.
```
It actually used to be called `coin` in an earlier version of the patch, but just s/coins/coin/ changes were such a substantial fraction of the commit that I backed out. Can be done later, I guess.

sipa commented on src/undo.h:
```diff
@@ -12,56 +12,72 @@
-        READWRITE(vprevout);
+    template <typename Stream>
+    void Unserialize(Stream& s) {
+        // TODO: avoid reimplementing vector deserializer
```
I have a plan for doing this cleanly, but it requires some refactoring in serialize.h which I'd like to avoid in this PR.

sipa commented on src/undo.h:
```diff
@@ -12,56 +12,72 @@
+        // TODO: avoid reimplementing vector deserializer
+        uint64_t count;
+        ::Unserialize(s, COMPACTSIZE(count));
+        if (count > 111111) { // TODO: avoid hardcoding max txouts per tx
```
Depending on consensus.h and deriving the constant directly. Alternatively, we may not need this protection if we'd use the std::vector deserialization code (see TODO above).

sipa committed:
Rename CCoinsCacheEntry::coins to coin

sipa commented:
Here is a graph of the speed and memory usage of a `-reindex-chainstate` until the default assumevalid point, with infinity -dbcache, on a 2.2GHz dual-cpu 28-core 56-threads machine with 256GiB RAM. Note that the x axis is progress (so proportional to the number of transactions).

![reindex](https://cloud.githubusercontent.com/assets/548488/25756823/0e61e73e-317d-11e7-9470-9f09db5ae131.png)

jnewbery commented on src/coins.h:
```diff
@@ -20,251 +20,90 @@
- *              * 00: special txout type pay-to-pubkey-hash
- *              * 8c988f1a4a4de2161e0f50aac7f17e7f9555caa4: address uint160
- *  - height = 120891
+ * - VARINT(coinbase + height * 2)
```
I found this quite confusing until I worked out you were packing bits into a single VARINT (partly because I didn't understand that coinbase was supposed to be 1 or 0). I don't know if this would be clearer as: `VARINT(height|coinbase)` ie height bits concatenated with coinbase bit.

sipa replied:
Rewritten in a different way.

jnewbery commented on src/coins.h:
```diff
@@ -336,97 +176,67 @@ class CCoinsViewBacked : public CCoinsView
 };
 
 
 class CCoinsViewCache;
```
now that you've removed CCoinsModifier, I believe you can remove this forward declaration of CCoinsViewCache.

sipa replied:
Fixed.

jnewbery reviewed (commented):
I've reviewed up to 4f09ae53d8cc3b570b04ea72b8f9844f93543823 "Introduce new per-txout CCoinsViewCache functions" and have a few nits. I'll finish reviewing next week.

sipa commented:
Same machine, but flushing at 1300 MiB:
![reindex-1300](https://cloud.githubusercontent.com/assets/548488/25769030/c84fe65e-31c4-11e7-8819-264c44e50ddf.png)

theuni commented on src/coins.cpp:
```diff
@@ -152,6 +153,55 @@ CCoinsModifier CCoinsViewCache::ModifyNewCoins(const uint256 &txid, bool coinbas
 
+void CCoinsViewCache::AddCoin(const COutPoint &outpoint, CCoin&& coin, bool possible_overwrite) {
+    assert(!coin.IsPruned());
+    auto ret = cacheCoins.emplace(std::piecewise_construct, std::forward_as_tuple(outpoint.hash), std::tuple<>());
```
std::tie would make this a little easier to read:
```c++
CCoinsMap::iterator it;
bool inserted;
std::tie(it, inserted) = cacheCoins.emplace(...
```

sipa replied:
Fixed.

sipa commented on src/coins.h:
```diff
@@ -20,251 +20,90 @@
 public:
-    //! whether transaction is a coinbase
-    bool fCoinBase;
+    //! unspent transaction outputs; spent outputs are .IsNull()
```
Fixed.

sipa commented on src/coins.h:
```diff
@@ -20,251 +20,90 @@
 
-    //! unspent transaction outputs; spent outputs are .IsNull(); spent outputs at the end of the array are dropped
-    std::vector<CTxOut> vout;
+    //! whether transaction is a coinbase
```
Fixed.

sipa commented on src/coins.h:
```diff
@@ -20,251 +20,90 @@
+    //! whether transaction is a coinbase
+    unsigned int fCoinBase : 1;
 
     //! at which height this transaction was included in the active block chain
```
Fixed.

sipa commented on src/coins.h:
```diff
@@ -20,251 +20,90 @@
-    CCoins(const CTransaction &tx, int nHeightIn) {
-        FromTx(tx, nHeightIn);
-    }
+    //! construct a CCoin from a CTransaction, at a given height
```
Fixed.

sipa commented on src/undo.h:
```diff
@@ -12,53 +12,70 @@
+        ::Unserialize(s, REF(CTxOutCompressor(REF(txout->out))));
     }
+
+    CTxInUndoSerializer(const CCoin* coins) : txout(const_cast<CCoin*>(coins)) {}
```
Fixed.

sipa commented on src/net_processing.cpp:
```diff
@@ -914,8 +914,7 @@ bool static AlreadyHave(const CInv& inv) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
                    mempool.exists(inv.hash) ||
-                   mapOrphanTransactions.count(inv.hash) ||
-                   pcoinsTip->HaveCoinsInCache(inv.hash);
+                   mapOrphanTransactions.count(inv.hash);
```
Fixed.

sipa commented on src/rpc/rawtransaction.cpp:
```diff
@@ -219,9 +219,13 @@ UniValue gettxoutproof(const JSONRPCRequest& request)
-            pblockindex = chainActive[coins.nHeight];
+        CCoin coins;
+        size_t o = 0;
+        while (pblockindex == nullptr && o < 1024) {
```
Fixed by introducing a helper function `AccessByTxid`.

sipa commented:
I've pushed a few more changes (rewriting the commits, but not a rebase), renaming some of the classes and variables to be more consistent.

For any further changes I'll create fixup commits in order to not hurt review.

ryanofsky commented:
utACK 37b273524984eb034f68277a635fa7265ece638e. Thanks for implementing so many review suggestions. Removing coins->Clear() from ApplyTxInUndo in an earlier commit, renaming `CCoin` to `Coin`, changing stats.nSerializedSize were the only significant changes since my last review that I don't remember seeing suggested, and these all look good.

sdaftuar commented on src/rpc/rawtransaction.cpp:
```diff
@@ -219,9 +219,8 @@ UniValue gettxoutproof(const JSONRPCRequest& request)
-        if (pcoinsTip->GetCoins(oneTxid, coins) && coins.nHeight > 0 && coins.nHeight <= chainActive.Height())
-            pblockindex = chainActive[coins.nHeight];
+        const Coin& coin = AccessByTxid(*pcoinsTip, oneTxid);
+        if (!coin.IsPruned()) pblockindex = chainActive[coin.nHeight];
```
It looks like the old code had sanity checking on the height, to prevent out-of-bounds access into chainActive.  I assume we should put that protection back in?

sipa replied:
Fixed.

sdaftuar commented on src/txmempool.cpp:
```diff
@@ -1048,7 +1051,7 @@ void CTxMemPool::TrimToSize(size_t sizelimit, std::vector<uint256>* pvNoSpendsRe
                     auto iter = mapNextTx.lower_bound(COutPoint(txin.prevout.hash, 0));
                     if (iter == mapNextTx.end() || iter->first->hash != txin.prevout.hash)
-                        pvNoSpendsRemaining->push_back(txin.prevout.hash);
+                        pvNoSpendsRemaining->push_back(txin.prevout);
```
Shouldn't this be including all outpoints that are no longer spent, rather than just outpoints from transactions where none of the outpoints are spent?

sipa replied:
Nice catch, fixed.

sdaftuar replied:
The `if (!mapNextTx.count(txin.prevout))` guard isn't necessary here, right?

sdaftuar reviewed (commented):
I've spent a bunch of time code reviewing today; got up to 21791e4fbc0bf385283ed82ff736961d45b0d6d6.

Posting my current set of comments now; will pick review up again on Monday.

TheBlueMatt commented on src/txdb.h:
```diff
@@ -75,11 +73,13 @@ class CCoinsViewDB : public CCoinsView
 
-    bool GetCoins(const uint256 &txid, CCoins &coins) const;
-    bool HaveCoins(const uint256 &txid) const;
+    bool GetCoins(const COutPoint &outpoint, Coin &coin) const;
```
nit: while you're changing these lines, it'd be nice to add overrides where appropriate.

sipa replied:
Done.

sipa committed:
Report on-disk size in gettxoutsetinfo

TheBlueMatt reviewed (commented):
Got through a big chunk, but still a bunch to review.

sipa commented on src/txmempool.h:
```diff
@@ -587,7 +601,7 @@ class CTxMemPool
       *  pvNoSpendsRemaining, if set, will be populated with the list of transactions
```
fixed

sipa commented on src/coins.cpp:
```diff
@@ -152,6 +153,57 @@ CCoinsModifier CCoinsViewCache::ModifyNewCoins(const uint256 &txid, bool coinbas
+    const uint256& txid = tx.GetHash();
+    for (size_t i = 0; i < tx.vout.size(); ++i) {
+        if (!tx.vout[i].scriptPubKey.IsUnspendable()) {
+            cache.AddCoin(COutPoint(txid, i), Coin(tx.vout[i], nHeight, fCoinbase), fCoinbase);
```
fixed

sipa commented on src/rpc/blockchain.cpp:
```diff
@@ -778,12 +778,15 @@ static void ApplyStats(CCoinsStats &stats, CHashWriter& ss, const uint256& hash,
     assert(!outputs.empty());
     ss << hash;
     ss << VARINT(outputs.begin()->second.nHeight * 2 + outputs.begin()->second.fCoinBase);
+    stats.nSerializedSize += 30 + 4 * outputs.size(); // Account for 1/16 chance of txid duplication in LevelDB.
```
I tried to fix this by writing a big comment about the data format and the rationale for the formula, only to discover it was wrong. Then I went on to fix it, and take even more cases into account, but the result still was pretty far off from the actual disk size. Instead, I removed the field entirely and replaced it with a disk_size element that is computed by LevelDB directly (but is not guaranteed to be deterministic).

sipa commented on src/validation.cpp:
```diff
@@ -1490,46 +1490,42 @@ bool AbortNode(CValidationState& state, const std::string& strMessage, const std
+ * @return True when the UTXO state was consistent with the undo data.
  */
-bool ApplyTxInUndo(const CTxInUndo& undo, CCoinsViewCache& view, const COutPoint& out)
+DisconnectResult ApplyTxInUndo(const CTxInUndo& undo, CCoinsViewCache& view, const COutPoint& out)
```
Ugh, that means exposing DisconnectResult :(

sipa replied:
Fixed by changing the type to int :(

sipa commented on src/coins.h:
```diff
@@ -412,41 +413,32 @@ class CCoinsViewCache : public CCoinsViewBacked
-     * TODO: return a reference to a Coin after changing CCoinsViewCache storage.
      */
-    const Coin AccessCoin(const COutPoint &output) const;
+    const Coin& AccessCoin(const COutPoint &output) const;
```
As explained on IRC, references to std::unordered_map entries always remain valid, even under concurrent modification/insertion (including when a rehash is triggered), except when the entry itself is deleted. Iterators do get invalidated by rehashing, but CCoinsViewCache does not allow iterators to escape anymore (they used to be, inside CCoinsModifier).

sipa commented on src/rpc/blockchain.cpp:
```diff
@@ -773,6 +773,25 @@ struct CCoinsStats
+    assert(!outputs.empty());
+    ss << hash;
+    ss << VARINT(outputs.begin()->second.nHeight * 2 + outputs.begin()->second.fCoinBase);
+    stats.nSerializedSize += 30 + 4 * outputs.size(); // Account for 1/16 chance of txid duplication in LevelDB.
```
Deleted this whole thing (see my response to @sdaftuar's comment on the same line).

sipa commented on src/validation.cpp:
```diff
@@ -1276,11 +1276,8 @@ void UpdateCoins(const CTransaction& tx, CCoinsViewCache& inputs, CTxUndo &txund
                 assert(false);
             // mark an outpoint spent, and construct undo information
-            txundo.vprevout.push_back(CTxInUndo(coins->vout[nPos]));
+            txundo.vprevout.push_back(CCoin(std::move(coins->vout[nPos]), coins->nHeight, coins->fCoinBase));
```
Agree, but I'm not going to rewrite history to fix that now. I can remember it when squashing.

sipa replied:
Actually, that std::move doesn't exist anywhere in the current PR. I think you were commenting on an old commit.

sipa commented on src/coins.cpp:
```diff
@@ -152,6 +152,55 @@ CCoinsModifier CCoinsViewCache::ModifyNewCoins(const uint256 &txid, bool coinbas
+        if (ret.first->second.coins.IsAvailable(outpoint.n)) {
+            throw std::logic_error("Adding new coin that replaces non-pruned entry");
+        }
+        fresh = !(ret.first->second.flags & CCoinsCacheEntry::DIRTY);
```
Agree, but again I'd rather not rewrite history. I can fix it when squashing, but just always setting it to false would not work (it would break the tests, which are very useful in determining that the correct behaviour remains during the whole changeset).

sipa commented on src/coins.cpp:
```diff
@@ -9,169 +9,116 @@
-    CCoinsMap::const_iterator it = FetchCoins(txid);
+void CCoinsViewCache::SpendCoin(const COutPoint &outpoint, CCoin* moveout) {
+    CCoinsMap::iterator it = FetchCoins(outpoint);
+    if (it == cacheCoins.end()) return;
```
Meh.

sipa commented on src/validation.h:
```diff
@@ -450,8 +450,8 @@ class CScriptCheck
     CScriptCheck(): amount(0), ptxTo(0), nIn(0), nFlags(0), cacheStore(false), error(SCRIPT_ERR_UNKNOWN_ERROR) {}
-    CScriptCheck(const CCoins& txFromIn, const CTransaction& txToIn, unsigned int nInIn, unsigned int nFlagsIn, bool cacheIn, PrecomputedTransactionData* txdataIn) :
-        scriptPubKey(txFromIn.vout[txToIn.vin[nInIn].prevout.n].scriptPubKey), amount(txFromIn.vout[txToIn.vin[nInIn].prevout.n].nValue),
+    CScriptCheck(const Coin& txFromIn, const CTransaction& txToIn, unsigned int nInIn, unsigned int nFlagsIn, bool cacheIn, PrecomputedTransactionData* txdataIn) :
```
Sounds good. Will do when squashing.

sipa commented on src/undo.h:
```diff
@@ -12,53 +12,72 @@
+        ::Unserialize(s, REF(CTxOutCompressor(REF(txout->out))));
     }
+
+    CTxInUndoSerializer(const CCoin* coins) : txout(const_cast<CCoin*>(coins)) {}
```
Fixed.

sipa commented on src/undo.h:
```diff
@@ -12,53 +12,72 @@
+        // TODO: avoid reimplementing vector deserializer
+        uint64_t count;
+        ::Unserialize(s, COMPACTSIZE(count));
+        if (count > 111111) { // TODO: avoid hardcoding max txouts per tx
```
Fixed.

sipa commented on src/validation.cpp:
```diff
@@ -1564,13 +1564,12 @@ static bool DisconnectBlock(const CBlock& block, CValidationState& state, const
                 const COutPoint &out = tx.vin[j].prevout;
-                const CCoin &undo = txundo.vprevout[j];
-                if (!ApplyTxInUndo(undo, view, out))
+                if (!ApplyTxInUndo(std::move(txundo.vprevout[j]), view, out))
```
Fixed.

sipa commented on src/coins.cpp:
```diff
@@ -285,35 +227,21 @@ bool CCoinsViewCache::HaveInputs(const CTransaction& tx) const
-        // If the coin still exists after the modification, add the new usage
-        cache.cachedCoinsUsage += it->second.coins.DynamicMemoryUsage();
+    COutPoint iter(txid, 0);
+    while (iter.n * ::GetSerializeSize(CTxOut(), SER_NETWORK, PROTOCOL_VERSION) < MAX_BLOCK_BASE_SIZE) {
```
Done.

sipa commented on src/coins.h:
```diff
@@ -460,17 +280,16 @@ class CCoinsViewCache : public CCoinsViewBacked
     CCoinsViewCache(const CCoinsViewCache &);
 };
 
+//! Utility function to add all of a transaction's outputs to a cache.
```
Done.

sipa commented on src/rpc/rawtransaction.cpp:
```diff
@@ -839,9 +838,12 @@ UniValue sendrawtransaction(const JSONRPCRequest& request)
+    bool fHaveChain = false;
+    for (size_t o = 0; !fHaveChain && o < tx->vout.size(); o++) {
+        const Coin& existingCoin = view.AccessCoin(COutPoint(hashTx, o));
+        fHaveChain = !existingCoin.IsPruned() && existingCoin.nHeight < 1000000000;
```
Done.

sipa commented on src/bench/ccoins_caching.cpp:
```diff
@@ -35,14 +35,14 @@ SetupDummyInputs(CBasicKeyStore& keystoreRet, CCoinsViewCache& coinsRet)
     dummyTransactions[0].vout[1].nValue = 50 * CENT;
     dummyTransactions[0].vout[1].scriptPubKey << ToByteVector(key[1].GetPubKey()) << OP_CHECKSIG;
-    coinsRet.ModifyCoins(dummyTransactions[0].GetHash())->FromTx(dummyTransactions[0], 0);
+    AddCoins(coinsRet, dummyTransactions[0], 0);
```
Right, there is no equivalent of ModifyCoins anymore.

sipa commented on src/init.cpp:
```diff
@@ -146,9 +146,9 @@ class CCoinsViewErrorCatcher : public CCoinsViewBacked
 public:
     CCoinsViewErrorCatcher(CCoinsView* view) : CCoinsViewBacked(view) {}
-    bool GetCoins(const uint256 &txid, CCoins &coins) const {
+    bool GetCoins(const COutPoint &outpoint, Coin &coin) const {
```
Added a bunch of overrides.

sipa commented on src/coins.cpp:
```diff
@@ -152,6 +152,55 @@ CCoinsModifier CCoinsViewCache::ModifyNewCoins(const uint256 &txid, bool coinbas
+    if (it == cacheCoins.end()) return;
+    cachedCoinsUsage -= it->second.coins.DynamicMemoryUsage();
+    if (moveout && it->second.coins.IsAvailable(outpoint.n)) {
+        *moveout = CCoin(std::move(it->second.coins.vout[outpoint.n]), it->second.coins.nHeight, it->second.coins.fCoinBase);
```
Go to fix this in a squashed version.

sipa commented:
Squashed history with fixups. Original history:
* 83367a95f922f6a5aec8b755e55c748e2edcbbc6 Add SizeEstimate to CDBBatch
* ac6910b5f6fd7fcfc2b002afe4a69f9ebca499fe error() in disconnect for disk corruption, not inconsistency
* f7beab2b69b4e239e6c6d92062dd7a5533f7fd9c Introduce CHashVerifier to hash read data
* e3d9ac37e1f7b74df263f9fbff2a482169284afe Add specialization of SipHash for 256 + 32 bit data
* 5be8c484780f1e66ee6d2e089d40a4d702e8a2f1 Remove/ignore tx version in utxo and undo
* ec8c2615d00cefbdec4f9af6c68545f4a0814fe2 Store/allow tx metadata in all undo records
* ac70d1fc96737eb49d6ca6ba8adf8e24d30a7f7d Introduce Coin, a single unspent output
* 50a8df05a1849cf7fe85e14226725a3e91390b0b Replace CTxInUndo with Coin
* c3ce6c92f7502d075118ed013f482fbd55ec102a Optimization: Coin&& to ApplyTxInUndo
* 777f4c6b63bbb4ce8e3afe51c02696893f4b165a Introduce new per-txout CCoinsViewCache functions
* 8df47a3ae67f1ce270b7612a61c5a136bf75297e Switch from per-tx to per-txout CCoinsViewCache methods in some places
* 546681621bbe8086fd38b568d0fb95cab224087e Switch CScriptCheck to use Coin instead of CCoins
* ce86045f34eb87cf32d50cd2e73a267cf2073d73 Switch tests from ModifyCoins to AddCoin/SpendCoin
* b5144262ebdacee7fe0cb77bf19186a7a1bde497 Remove ModifyCoins/ModifyNewCoins
* 593a6324fbaf6bf1188c15ad23d670213e9cb41c Replace CCoins-based CTxMemPool::pruneSpent with isSpent
* deb42bb6c84e93d3f5e53c9007d58a7abdc45001 Refactor GetUTXOStats in preparation for per-COutPoint iteration
* 21791e4fbc0bf385283ed82ff736961d45b0d6d6 Switch CCoinsView and chainstate db from per-txid to per-txout
* 6fd519e2d1fad48fbe5bf5b335c6fb0606d906e5 Remove unused CCoins methods
* 575ffb39d6563c73b057690ab3171aa96097c27e Pack Coin more tightly
* f2cd835a18d3f784fe45fa0421120d8fb96a6dce Reduce reserved memory space for flushing
* b951db2885c97c3bc71b58aa00a3a897b75a8340 Upgrade from per-tx database to per-txout
* 229b89c36f104a7f26a2d40b604e631cc1c56814 [MOVEONLY] Move old CCoins class to txdb.cpp
* fdab87c96cbd4e92aec40fc10ba83dd2ef2ddbe8 Merge CCoinsViewCache's GetOutputFor and AccessCoin
* 37b273524984eb034f68277a635fa7265ece638e Rename CCoinsCacheEntry::coins to coin
* e581f28c0f7c7bba36f28b1c2a69c7a584e54edb f Replace CTxInUndo with Coin: initialize count variale
* 69f14c4c7be6fd233a420953255a21119e11096f f Introduce new per-txout CCoinsViewCache functions: add comment explaining fCoinbase
* 71ccd9d1bd50a523087ef74140af2032907f7656 f Switch CCoinsView and chainstate db from per-txid to per-txout: fix txmempool comment
* fdfafc54503aca8645c9bd8816f590833debf363 f Remove/ignore tx version in utxo and undo: remove ill-defined bytes_serialized
* 9da167ec95b3bbf44ca545adcc0855d8d0c25642 Report on-disk size in gettxoutsetinfo
* 3e04e32a10df48965981e86c559cf4b8691e7699 f Switch CCoinsView and chainstate db from per-txid to per-txout: re-add sanity check in gettxoutproof
* c8985414f376e73ae301d282b6e38e8accf493f9 f Switch CCoinsView and chainstate db from per-txid to per-txout: CTxMempool::TrimToSize should not use per-txid logic for pvNoRemainingSpends
* b69f3bc3400e6693aa04b00c054f5018b6ffa954 f Replace CTxInUndo with Coin: split TxInUndoSerializer off from deserializer
* 544c34b202725c932cb9edbd8be655bc7764fa85 f Replace CTxInUndo with Coin: don't hardcode max outputs per block
* bf4dadbd98d07a07b87fea577ef9d38328d8607d f error() in disconnect for disk corruption, not inconsistency: correct type in extern ApplyTxInUndo in coins_tests.cpp
* 6ac7ea345d18d13bac4bc00e513eb3065a36c713 f Optimization: Coin&& to ApplyTxInUndo: Comment that all of txundo.prevout was moved
* 369afe4208ede1741cbde553364a1fcca7d44781 f Introduce new per-txout CCoinsViewCache functions: do not compute CTxOut() size in loop
* dc2bb20a396317e47e17bb0533ca0e42d3f81fb1 f Introduce new per-txout CCoinsViewCache functions: Add comment explaining AddCoins potential_overwrite assumptions
* 5611b76c963d02bd267127122222db85a5e3761a f Switch from per-tx to per-txout CCoinsViewCache methods in some places: remove redundant nHeight < 1000000000 logic
* 51c7c4235f12de8271ff2ed0447571b6a3e71a23 f Switch CCoinsView and chainstate db from per-txid to per-txout: add overrides
* b5621d6dbd41dfde331afd45ab6e0c4f8c8c2b82 f Switch CCoinsView and chainstate db from per-txid to per-txout: still try COutPoint(hash, 0) in AlreadyHave
* ea2a5b937a625f738c97314de48e117be9e76abb Only pass things committed to by tx's witness hash to CScriptCheck

laanwj commented on src/rpc/blockchain.cpp:
```diff
@@ -878,8 +891,8 @@ UniValue gettxoutsetinfo(const JSONRPCRequest& request)
             "  \"transactions\": n,      (numeric) The number of transactions\n"
             "  \"txouts\": n,            (numeric) The number of output transactions\n"
-            "  \"bytes_serialized\": n,  (numeric) The serialized size\n"
             "  \"hash_serialized\": \"hash\",   (string) The serialized hash\n"
```
doc: `hash_serialized_2`

laanwj commented:
I've been testing this on a slow system: on an ARM (i.MX6), USB2 external harddisk, the database upgrade took about 50 minutes:
```
2017-05-15 10:27:45 Upgrading database...
2017-05-15 11:20:25 LoadBlockIndexDB: last block file = 869
```
Not necessarily in this pull, but some kind of progress feedback like for rescans / reindexes would be indispensable. Users might assume that the client has crashed after no log messages for such a long time.

sdaftuar reviewed (commented):
Code review ACK apart from the tests, which I'm still reviewing/testing.

The database upgrade took about 5 minutes on my machine (chainstate stored on SSD).

sipa commented:
> Not necessarily in this pull, but some kind of progress feedback like for rescans / reindexes would be indispensable. Users might assume that the client has crashed after no log messages for such a long time.

Yes, fully agree. There would need to be some UI interaction in the GUI, and at least some progress reported in debug.log during the process. I'd prefer to do that after merging, though.

ryanofsky commented on test/functional/blockchain.py:
```diff
@@ -49,6 +49,9 @@ def _test_gettxoutsetinfo(self):
         assert_equal(res['height'], 200)
         assert_equal(res['txouts'], 200)
+        size = res['disk_size']
+        assert size > 6400
```
In commit "Report on-disk size in gettxoutsetinfo"

Could assert `6400 < size < 64000`

ryanofsky reviewed (commented):
utACK a86f1c959c4beaf2b2667b000983f22de6acadf2

(Changes since last review: added fixes 37b273524984eb034f68277a635fa7265ece638e..ea2a5b937a625f738c97314de48e117be9e76abb, squashed ea2a5b937a625f738c97314de48e117be9e76abb -> 02aa1bec3e4e2b15caa3bd233c72ce1449133a2f, and merged new master 94e52273f30fc9f3f1a7b58778ed21781bb2a744.)

sdaftuar commented on src/coins.cpp:
```diff
@@ -262,21 +214,14 @@ unsigned int CCoinsViewCache::GetCacheSize() const {
     CAmount nResult = 0;
     for (unsigned int i = 0; i < tx.vin.size(); i++)
-        nResult += GetOutputFor(tx.vin[i]).nValue;
+        nResult += AccessCoin(tx.vin[i].prevout).out.nValue;
```
I was wondering what happens in this function if an input doesn't exist -- in the old code, it appears we'd `assert()` in `GetOutputFor()`, whereas in the new code we'll just silently return an incorrect value (because nValue is -1 in that case).  Do you think it's worth asserting here that AccessCoin is returning something valid?

sipa replied:
I can add an assert if you're concerned, but given that this code hasn't asserted probably ever, it should be fine. I really want to get rid of this method, and instead make the validation code build a vector of CCoin& references for all its inputs, and use that everywhere, as opposed to looking things up several times inline.

sdaftuar replied:
Your plan sounds good!

sdaftuar commented:
ACK a86f1c959c4beaf2b2667b000983f22de6acadf2, this is awesome.

Though it looks like this needs rebase now.  (Also @sipa I'm confused about why you included the merge commit in this PR?)

I did some benchmarking of validation speed in my simulation environment, comparing master to this PR.  This makes better use of pcoinsTip, which means we flush less often and thus expect a big performance win from that effect alone (which I did observe).  Moreover, even if I compare master against this PR using a -dbcache big enough that neither one flushes to disk during the simulation period, this PR still performs substantially better -- 33% improvement in the "Connect transactions" time in ConnectBlock, and roughly 8% reduction overall in "Connect block" time.

sipa commented:
@sdaftuar Thanks, rebased. The merge commit at the end was to avoid a CI error that was fixed in master, but existed at the time the code branched off.

sdaftuar commented:
re-ACK 71837d800cdacd92ddb5bf8536c0e16338a6a889

gmaxwell commented on src/coins.cpp:
```diff
@@ -4,174 +4,127 @@
+    it->second.coin = std::move(coin);
+    it->second.flags |= CCoinsCacheEntry::DIRTY | (fresh ? CCoinsCacheEntry::FRESH : 0);
+    cachedCoinsUsage += it->second.coin.DynamicMemoryUsage();
+}
```
Having trouble with the logic around !inserted here. If an entry already existed, we remove its size. (potentially making cachedCoinsUsage underflow). Then if we were not willing to overwrite, we throw leaving the accounting in an odd state.

Why is the logic here not just having an if(inserted) on the last line and elimiating the !inserted block?

sipa replied:
> If an entry already existed, we remove its size. (potentially making cachedCoinsUsage underflow).

Underflow is not possible if the cachedCoinsUsage value was consistent with the database.

> Then if we were not willing to overwrite, we throw leaving the accounting in an odd state.

That throw should be regarded as an assertion failure, except for the fact that tests can catch it.

> Why is the logic here not just having an if(inserted) on the last line and elimiating the !inserted block?

It's possible the entry was overwritten with a different value, needing correction.

gmaxwell commented on src/coins.h:
```diff
@@ -460,17 +272,22 @@ class CCoinsViewCache : public CCoinsViewBacked
+//! Utility function to add all of a transaction's outputs to a cache.
+// It assumes that overwrites are only possible for coinbase transactions,
+// TODO: pass in a boolean to limit these possible overwrites to pre-BIP34
+// coinbases.
```
preferably it should limit it to just the two overwrites that re actually needed.

gmaxwell commented on src/net_processing.cpp:
```diff
@@ -911,12 +911,10 @@ bool static AlreadyHave(const CInv& inv) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
                    mempool.exists(inv.hash) ||
                    mapOrphanTransactions.count(inv.hash) ||
-                   pcoinsTip->HaveCoinsInCache(inv.hash);
+                   pcoinsTip->HaveCoinsInCache(COutPoint(inv.hash, 0)); // Best effort: only try output 0
```
It may be advisable to also check 1.  The reason is that recently confirmed transactions will get requested again, and we're currently protected by this check. But if the transaction chained off output 0 it won't be in the cache anymore. Also testing 1 would be a lot less often that we would erroneously double request a transaction we just confirmed.

gmaxwell commented on src/txdb.cpp:
```diff
@@ -63,8 +88,14 @@ bool CCoinsViewDB::BatchWrite(CCoinsMap &mapCoins, const uint256 &hashBlock) {
-    LogPrint(BCLog::COINDB, "Committing %u changed transactions (out of %u) to coin database...\n", (unsigned int)changed, (unsigned int)count);
-    return db.WriteBatch(batch);
+    bool ret = db.WriteBatch(batch);
+    LogPrint(BCLog::COINDB, "Committed %u changed transactions (out of %u) to coin database...\n", (unsigned int)changed, (unsigned int)count);
```
changed txouts?

gmaxwell commented:
I tested the tests by manually injecting faults in coins.cpp, here are my findings:

CCoinsViewCache::FetchCoins
commenting out cachedCoinsUsage += ret->second.coin.DynamicMemoryUsage(); still passes coins_tests.

CCoinsViewCache::AddCoin
inverting condition in if (!inserted) passes coins_tests.

CCoinsViewCache::AddCoin
commenting out cachedCoinsUsage += it->  passes coins_tests.

AddCoins
Adding 1 || to the !tx.[...].IsUnspendtable() condition passes coins_tests.

CCoinsViewCache::SpendCoin
commenting out cachedCoinsUsage -= it->seco  passes coins_tests

CCoinsViewCache::HaveCoins
removing && !it->second.coin.IsPruned() passes coins_tests

CCoinsViewCache::BatchWrite
commenting out  cachedCoinsUsage += entry.coin.DynamicMemoryUsag passes coins_tests
(anywhere in the function)

CCoinsViewCache::BatchWrite
commenting out hashBlock = hashBlockIn passes coins_tests

CCoinsViewCache::Flush
commenting out cachedCoinsUsage = 0; passes coins_tests

CCoinsViewCache::Uncache
commenting out  cachedCoinsUsage += entry.coin.DynamicMemoryUsag passes coins_tests

CCoinsViewCache::Uncache
adding return to top of function passes coins_test

CCoinsViewCache::GetCacheSize
changing function to return 0 passes coins_test

CCoinsViewCache::GetValueIn
changing function to return 0 passes coins_test

CCoinsViewCache::HaveInputs
changing function to return true; passes coins test

AccessByTxid
changing function to return coinEmpty; passes coins test

sipa commented:
@gmaxwell I've added a commit that extends the coins_test to (hopefully) address all the cases you find, except for HaveInputs and GetValueIn (which I plan to remove in a follow-up PR).

gmaxwell commented:
AddCoins
Adding 1 || to the !tx.[...].IsUnspendtable() condition passes coins_tests.

^ that isn't fixed yet.

sipa commented:
@gmaxwell Fixed.

gmaxwell commented on src/undo.h:
```diff
@@ -7,58 +7,86 @@
+        // TODO: avoid reimplementing vector deserializer
+        uint64_t count = 0;
+        ::Unserialize(s, COMPACTSIZE(count));
+        if (count > MAX_OUTPUTS_PER_BLOCK) {
```
Isn't this the maximum here the number of _INPUTS_ consumed by a block?  (e.g. base size divided by 40?)  I suppose the output number is higher so this at least won't cause failures, but I think it's wrong.

sipa replied:
Indeed! Fixed.

sipa committed:
Extend coins_tests

sipa committed:
Increase travis unit test timeout

gmaxwell commented:
$ ./src/bitcoin-cli -testnet=1 gettxoutsetinfo
{
  "height": 1125236,
  "bestblock":
"00000000000002019cb4550f3046934d4355a56451eefc66b2023d5152b1d1ee",
  "transactions": 4174348,
  "txouts": 15243146,
  "hash_serialized_2":
  "f091cba43924f4c0561740f85448cb7b3a5b383174a65087793a5fd2a2835bd3",
  "disk_size": 0,   <<<< interaction with leveldb is slightly surprising
  "total_amount": 20460499.61149857
}

gmaxwell commented:
sync testnet. restart.

invalidate block 5. 

{
  "height": 4,
  "bestblock":
"000000008b5d0af9ffb1741e38b17b193bd12d7683401cecd2fd94f548b6e5dd",
  "transactions": 4,
  "txouts": 4,
  "hash_serialized_2":
  "9ea56b9590b2ce5582a0013e2bfd08222334247ec2bdbe4c71dabd7210040b71",
  "disk_size": 808430596,
  "total_amount": 200.00000000
}

restart

{
  "height": 4,
  "bestblock": "000000008b5d0af9ffb1741e38b17b193bd12d7683401cecd2fd94f548b6e5dd",
  "transactions": 4,
  "txouts": 4,
  "hash_serialized_2": "9ea56b9590b2ce5582a0013e2bfd08222334247ec2bdbe4c71dabd7210040b71",
  "disk_size": 1199640169,
  "total_amount": 200.00000000
}

after another restart
  "disk_size": 561,

leveldb be crazy, yo. I doubt think it's wrong, but we should perhaps expect confused questions from users.

And gettxoutsetinfo crashes with an assert on slice validity in leveldb if called when the tip is height 0 (either having never synced any blocks, or after invalidating block 1) -- pretty sure this is not desired. :)

gmaxwell commented:
@sipa  it isn't a new bug introduced by this PR, but my ACK is waiting on a fix for the gettxoutsetinfo crash on an empty chainstate.

sdaftuar commented:
> @sipa it isn't a new bug introduced by this PR, but my ACK is waiting on a fix for the gettxoutsetinfo crash on an empty chainstate.

This seems to fix the crash @gmaxwell discovered (which is easy to reproduce in regtest), though I'm not sure if this is exactly the right fix:

```
diff --git a/src/txdb.cpp b/src/txdb.cpp
index 01108beae..d8d754617 100644
--- a/src/txdb.cpp
+++ b/src/txdb.cpp
@@ -128,10 +128,12 @@ CCoinsViewCursor *CCoinsViewDB::Cursor() const
        only need read operations on it, use a const-cast to get around
        that restriction.  */
     i->pcursor->Seek(DB_COIN);
-    // Cache key of first record
-    CoinsEntry entry(&i->keyTmp.second);
-    i->pcursor->GetKey(entry);
-    i->keyTmp.first = entry.key;
+    if (i->pcursor->Valid()) {
+        // Cache key of first record
+        CoinsEntry entry(&i->keyTmp.second);
+        i->pcursor->GetKey(entry);
+        i->keyTmp.first = entry.key;
+    }
     return i;
 }
```

ryanofsky commented on src/test/coins_tests.cpp:
```diff
@@ -133,36 +147,62 @@ BOOST_AUTO_TEST_CASE(coins_cache_simulation_test)
+                newcoin.out.nValue = insecure_rand();
+                newcoin.nHeight = 1;
+            }
+            if (newcoin.IsPruned()) {
```
In commit "Extend coins_tests"

I think it would be clearer to combine this if/else block with the previous one, so the different cases perform steps in a uniform order, and there are fewer conditions that need to be checked. The following seems to do the same work and drops around 10 lines of code:

```
if (insecure_rand() % 5 == 0 || coin.IsPruned()) {
    newcoin.out.nValue = insecure_rand();
    newcoin.nHeight = 1;
    if (insecure_rand() % 16 == 0 && coin.IsPruned()) {
        newcoin.out.scriptPubKey.assign(1 + (insecure_rand() & 0x3F), OP_RETURN);
        BOOST_CHECK(newcoin.out.scriptPubKey.IsUnspendable());
        added_an_unspendable_entry = true;
    } else {
        newcoin.out.scriptPubKey.assign(insecure_rand() & 0x3F, 0); // Random sizes so we can test memory usage accounting
        (coin.IsPruned() ? added_an_entry : updated_an_entry) = true;
        coin = newcoin;
    }
    stack.back()->AddCoin(COutPoint(txid, 0), std::move(newcoin), !coin.IsPruned() || insecure_rand() & 1);
} else {
    removed_an_entry = true;
    coin.Clear();
    stack.back()->SpendCoin(COutPoint(txid, 0));
}
```

ryanofsky commented on src/test/coins_tests.cpp:
```diff
@@ -133,36 +147,62 @@ BOOST_AUTO_TEST_CASE(coins_cache_simulation_test)
-                if (coins.IsPruned()) {
-                    added_an_entry = true;
+            Coin& coin = result[COutPoint(txid, 0)];
+            Coin newcoin;
```
"In commit "Extend coins_tests"

Maybe declare newcoin below closer to its first use.

ryanofsky reviewed (commented):
Obviously there are other issues still being worked out, but posting an updated utACK for 8ec6ed0bc964a3beb0ba6b8945b3219d6d8189b2. Changes since my previous review were a rebase that resolved conflicts with #8329, and 5 additional commits.

sipa commented:
Modified the last few commits to include @sdaftuar and @ryanofsky's suggestions.

gmaxwell reviewed (approved):
ACK. Great work

dougstrong77 commented:
Need helpwith all this

sdaftuar commented:
re-ACK 99eecb9

sipa commented:
Rebased on top of #10445, and squashed fixes into appropriate commits.

I think this is ready.

gmaxwell commented:
re-ACK

sdaftuar commented:
re-ACK

jnewbery commented on src/coins.h:
```diff
@@ -20,251 +20,90 @@
-
-    //! check whether the entire CCoins is spent
-    //! note that only !IsPruned() CCoins can be serialized
     bool IsPruned() const {
```
to me, `IsPruned()` makes less sense in the context of a single coin. For CCoins, `IsPruned()` meant that all the coins were spent and so the containing object could be pruned. For a single Coin class, I think `IsSpent()` makes more sense.

sipa replied:
Fixed using scripted-diff.

jnewbery commented on src/test/coins_tests.cpp:
```diff
@@ -242,18 +273,18 @@ BOOST_AUTO_TEST_CASE(updatecoins_simulation_test)
-    std::set<uint256> disconnectedids;
-    std::set<uint256> duplicateids;
-    std::set<uint256> utxoset;
+    std::set<COutPoint> coinbaseids;
```
rename to `coinbasecoins`? Same for next two variables.

sipa replied:
Fixed using scripted-diff.

jnewbery commented on src/test/coins_tests.cpp:
```diff
@@ -264,138 +295,144 @@ BOOST_AUTO_TEST_CASE(updatecoins_simulation_test)
+            tx.vout[0].scriptPubKey.assign(insecure_rand() & 0x3F, 0); // Random sizes so we can test memory usage accounting
             unsigned int height = insecure_rand();
-            CCoins oldcoins;
+            Coin oldcoins;
```
rename to `oldcoin`?

sipa replied:
Fixed using scripted-diff.

jnewbery commented on src/test/coins_tests.cpp:
```diff
@@ -264,138 +295,144 @@ BOOST_AUTO_TEST_CASE(updatecoins_simulation_test)
                 // The test is designed to ensure spending a duplicate coinbase will work properly
                 // if that ever happens and not resurrect the previously overwritten coinbase
-                if (duplicateids.count(prevouthash))
+                if (duplicateids.count(prevout))
```
braces, same line, or ternary operator

sipa replied:
Fixed.

jnewbery commented on src/test/coins_tests.cpp:
```diff
@@ -264,138 +295,144 @@ BOOST_AUTO_TEST_CASE(updatecoins_simulation_test)
         }
 
         //1/20 times undo a previous transaction
         else if (utxoset.size()) {
```
braces or same line

sipa replied:
Fixed.

jnewbery commented on src/test/coins_tests.cpp:
```diff
@@ -264,138 +295,144 @@ BOOST_AUTO_TEST_CASE(updatecoins_simulation_test)
-            uint256 undohash = tx.GetHash();
+            CTransaction &tx = std::get<0>(utxod->second);
+            CTxUndo &undo = std::get<1>(utxod->second);
+            Coin &origcoins = std::get<2>(utxod->second);
```
rename to `origcoin`?

sipa replied:
Fixed using scripted-diff.

jnewbery commented on src/rpc/blockchain.cpp:
```diff
@@ -963,37 +976,34 @@ UniValue gettxout(const JSONRPCRequest& request)
-        mempool.pruneSpent(hash, coins); // TODO: this should be done by the CCoinsViewMemPool
     } else {
-        if (!pcoinsTip->GetCoins(hash, coins))
+        if (!pcoinsTip->GetCoins(out, coin))
```
braces or same line

sipa replied:
Fixed.

jnewbery commented on src/rpc/blockchain.cpp:
```diff
@@ -963,37 +976,34 @@ UniValue gettxout(const JSONRPCRequest& request)
     CBlockIndex *pindex = it->second;
     ret.push_back(Pair("bestblock", pindex->GetBlockHash().GetHex()));
-    if ((unsigned int)coins.nHeight == MEMPOOL_HEIGHT)
+    if ((unsigned int)coin.nHeight == MEMPOOL_HEIGHT)
```
braces

sipa replied:
Fixed.

jnewbery commented on src/txmempool.h:
```diff
@@ -30,7 +30,7 @@
 
-/** Fake height value used in CCoins to signify they are only in the memory pool (since 0.8) */
+/** Fake height value used in Coin to signify they are only in the memory pool (since 0.8) */
 static const unsigned int MEMPOOL_HEIGHT = 0x7FFFFFFF;
```
Why not make this a uint32_t to match `Coin.nHeight`?

sipa replied:
Fixed.

jnewbery reviewed (commented):
I've got as far as 0dcdbbd0b36298299d59f3960cd8b0be48dd3b80
 - Switch CCoinsView and chainstate db from per-txid to per-txout

All of my comments are nits, mostly around naming where the names are still ___coins, when in fact the variable or function is now referring to single coin object. You can take them or leave them (or we can address in a future PR). They certainly shouldn't hold up merging this PR.

sipa committed:
scripted-diff: various renames for per-utxo consistency

Thanks to John Newberry for pointing these out.

-BEGIN VERIFY SCRIPT-
sed -i 's/\<GetCoins\>/GetCoin/g' src/*.cpp src/*.h src/*/*.cpp src/*/*.h
sed -i 's/\<HaveCoins\>/HaveCoin/g' src/*.cpp src/*.h src/*/*.cpp src/*/*.h
sed -i 's/\<HaveCoinsInCache\>/HaveCoinInCache/g' src/*.cpp src/*.h src/*/*.cpp src/*/*.h
sed -i 's/\<IsPruned\>/IsSpent/g' src/*.cpp src/*.h src/*/*.cpp src/*/*.h
sed -i 's/\<FetchCoins\>/FetchCoin/g' src/*.cpp src/*.h src/*/*.cpp src/*/*.h
sed -i 's/\<CoinsEntry\>/CoinEntry/g' src/*.cpp src/*.h src/*/*.cpp src/*/*.h
sed -i 's/\<vHashTxnToUncache\>/coins_to_uncache/g' src/*.cpp src/*.h src/*/*.cpp src/*/*.h
sed -i 's/\<vHashTxToUncache\>/coins_to_uncache/g' src/*.cpp src/*.h src/*/*.cpp src/*/*.h
sed -i 's/\<fHadTxInCache\>/had_coin_in_cache/g' src/*.cpp src/*.h src/*/*.cpp src/*/*.h
sed -i 's/\<coinbaseids\>/coinbase_coins/g' src/*.cpp src/*.h src/*/*.cpp src/*/*.h
sed -i 's/\<disconnectedids\>/disconnected_coins/g' src/*.cpp src/*.h src/*/*.cpp src/*/*.h
sed -i 's/\<duplicateids\>/duplicate_coins/g' src/*.cpp src/*.h src/*/*.cpp src/*/*.h
sed -i 's/\<oldcoins\>/old_coin/g' src/test/coins_tests.cpp
sed -i 's/\<origcoins\>/orig_coin/g' src/*.cpp src/*.h src/*/*.cpp src/*/*.h
-END VERIFY SCRIPT-

sipa commented on src/rpc/blockchain.cpp:
```diff
@@ -963,37 +976,34 @@ UniValue gettxout(const JSONRPCRequest& request)
         LOCK(mempool.cs);
         CCoinsViewMemPool view(pcoinsTip, mempool);
-        if (!view.GetCoins(hash, coins))
+        if (!view.GetCoins(out, coin) || mempool.isSpent(out)) // TODO: this should be done by the CCoinsViewMemPool
```
Fixed. The comment is not outdated.

jnewbery replied:
In that case, I don't understand the comment. I thought it meant that the CCoinsViewMemPool class should be responsible for pruning the spent funds, instead of only when called through the rpc or rest.

sipa commented on src/undo.h:
```diff
@@ -7,58 +7,86 @@
- *  last output of the affected transaction, its metadata as well
- *  (coinbase or not, height, transaction version)
+ *  Contains the prevout's CTxOut being spent, and its metadata as well
+ *  (coinbase or not, height). Earlier versions also stored the transaction
```
Done.

sipa commented on src/undo.h:
```diff
@@ -7,58 +7,86 @@
-            ::Serialize(s, VARINT(this->nVersion));
-        ::Serialize(s, CTxOutCompressor(REF(txout)));
+        ::Serialize(s, VARINT(txout->nHeight * 2 + (txout->fCoinBase ? 1 : 0)));
+        if (txout->nHeight > 0) {
```
Added a comment to explain.

sipa commented on src/undo.h:
```diff
@@ -7,58 +7,86 @@
-        ::Serialize(s, CTxOutCompressor(REF(txout)));
+        ::Serialize(s, VARINT(txout->nHeight * 2 + (txout->fCoinBase ? 1 : 0)));
+        if (txout->nHeight > 0) {
+            // Store a dummy for the transaction version.
```
Fixed.

sipa commented on test/functional/blockchain.py:
```diff
@@ -49,10 +49,12 @@ def _test_gettxoutsetinfo(self):
-        assert_equal(res['bytes_serialized'], 13924),
         assert_equal(res['bestblock'], node.getblockhash(200))
+        size = res['disk_size']
+        assert size > 6400
```
Will do, but this should go in #10396.

sipa commented on src/coins.h:
```diff
@@ -20,251 +20,90 @@
-        Cleanup();
+        uint32_t code = 0;
+        ::Unserialize(s, VARINT(code));
+        nHeight = code >> 1;
```
Not really... just inconsistently applying the best practices of using the faster operation (and not relying on the compiler optimizer for doing so).

sipa commented on src/coins.h:
```diff
@@ -306,17 +146,17 @@ class CCoinsViewCursor
-    //! Retrieve the CCoins (unspent transaction outputs) for a given txid
-    virtual bool GetCoins(const uint256 &txid, CCoins &coins) const;
+    //! Retrieve the Coin (unspent transaction output) for a given outpoint.
+    virtual bool GetCoins(const COutPoint &outpoint, Coin &coin) const;
```
Fixed using scripted-diff.

sipa commented on src/coins.h:
```diff
@@ -336,97 +179,66 @@ class CCoinsViewBacked : public CCoinsView
      * the backing CCoinsView are made.
      */
-    bool HaveCoinsInCache(const uint256 &txid) const;
+    bool HaveCoinsInCache(const COutPoint &outpoint) const;
```
Fixed using scripted-diff.

sipa commented on src/coins.h:
```diff
@@ -460,17 +272,22 @@ class CCoinsViewCache : public CCoinsViewBacked
-
 private:
-    CCoinsMap::const_iterator FetchCoins(const uint256 &txid) const;
+    CCoinsMap::iterator FetchCoins(const COutPoint &outpoint) const;
```
Fixed using scripted-diff.

sipa commented on src/txdb.cpp:
```diff
@@ -24,17 +25,40 @@ static const char DB_FLAG = 'F';
 
+namespace {
+
+struct CoinsEntry {
```
Fixed using scripted-diff.

sipa commented on src/validation.cpp:
```diff
@@ -344,7 +344,7 @@ static bool IsCurrentForFeeEstimation()
 bool AcceptToMemoryPoolWorker(CTxMemPool& pool, CValidationState& state, const CTransactionRef& ptx, bool fLimitFree,
                               bool* pfMissingInputs, int64_t nAcceptTime, std::list<CTransactionRef>* plTxnReplaced,
-                              bool fOverrideMempoolLimit, const CAmount& nAbsurdFee, std::vector<uint256>& vHashTxnToUncache)
+                              bool fOverrideMempoolLimit, const CAmount& nAbsurdFee, std::vector<COutPoint>& vHashTxnToUncache)
```
Fixed using scripted-diff.

sipa commented on src/validation.cpp:
```diff
@@ -437,30 +437,28 @@ bool AcceptToMemoryPoolWorker(CTxMemPool& pool, CValidationState& state, const C
-            return state.Invalid(false, REJECT_ALREADY_KNOWN, "txn-already-known");
+        for (size_t out = 0; out < tx.vout.size(); out++) {
+            COutPoint outpoint(hash, out);
+            bool fHadTxInCache = pcoinsTip->HaveCoinsInCache(outpoint);
```
Fixed using scripted-diff.

sipa commented on src/validation.cpp:
```diff
@@ -437,30 +437,28 @@ bool AcceptToMemoryPoolWorker(CTxMemPool& pool, CValidationState& state, const C
-            if (!pcoinsTip->HaveCoinsInCache(txin.prevout.hash))
-                vHashTxnToUncache.push_back(txin.prevout.hash);
-            if (!view.HaveCoins(txin.prevout.hash)) {
+            if (!pcoinsTip->HaveCoinsInCache(txin.prevout))
```
Fixed.

sipa commented on src/validation.cpp:
```diff
@@ -763,10 +761,10 @@ bool AcceptToMemoryPoolWithTime(CTxMemPool& pool, CValidationState &state, const
                         bool fOverrideMempoolLimit, const CAmount nAbsurdFee)
 {
-    std::vector<uint256> vHashTxToUncache;
+    std::vector<COutPoint> vHashTxToUncache;
```
Fixed using scripted-diff.

sipa commented on src/test/test_bitcoin_fuzzy.cpp:
```diff
@@ -168,7 +168,7 @@ int do_fuzz()
             try
             {
-                CCoins block;
+                Coin block;
```
Fixed.

jnewbery commented on src/rpc/blockchain.cpp:
```diff
@@ -963,37 +976,37 @@ UniValue gettxout(const JSONRPCRequest& request)
     CBlockIndex *pindex = it->second;
     ret.push_back(Pair("bestblock", pindex->GetBlockHash().GetHex()));
-    if ((unsigned int)coins.nHeight == MEMPOOL_HEIGHT)
+    if ((unsigned int)coin.nHeight == MEMPOOL_HEIGHT) {
```
cast not required. `coin.nHeight` and `MEMPOOL_HEIGHT` are both uint32_t

jnewbery reviewed (commented):
Finished reviewing. Just a few more nits.

I'm not familiar enough with the code to give a full ACK for functionality and non-regression, but I've reviewed as best I can for correctness and style and it looks good to me.

I've also done some limited test locally and haven't hit any problems.

sipa commented on src/init.cpp:
```diff
@@ -1455,6 +1455,11 @@ bool AppInitMain(boost::thread_group& threadGroup, CScheduler& scheduler)
                     if (fPruneMode)
                         CleanupBlockRevFiles();
+                } else {
+                    if (!pcoinsdbview->Upgrade()) {
```
Added some comments. The details about upgrading are in the cpp file, because I expect that the Upgrade method will remain, regardless of what upgrades it supports.

sipa commented on src/txdb.cpp:
```diff
@@ -215,3 +252,121 @@ bool CBlockTreeDB::LoadBlockIndexGuts(std::function<CBlockIndex*(const uint256&)
+
+namespace {
+
+class CCoins
```
Done.

sipa commented on src/txdb.cpp:
```diff
@@ -215,3 +252,121 @@ bool CBlockTreeDB::LoadBlockIndexGuts(std::function<CBlockIndex*(const uint256&)
+     * each bit in the bitmask represents the availability of one output, but the
+     * availabilities of the first two outputs are encoded separately
+     */
+    void CalcMaskSize(unsigned int &nBytes, unsigned int &nNonzeroBytes) const {
```
Great catch, done!

sipa commented on src/txdb.cpp:
```diff
@@ -215,3 +252,121 @@ bool CBlockTreeDB::LoadBlockIndexGuts(std::function<CBlockIndex*(const uint256&)
+        boost::this_thread::interruption_point();
+        std::pair<unsigned char, uint256> key;
+        if (pcursor->GetKey(key) && key.first == DB_COINS) {
+            CCoins old_coin;
```
Reverted.

jnewbery commented:
utACK https://github.com/bitcoin/bitcoin/pull/10195/commits/dfdf1d5fef732d24602e693f917e41ed357d4937

laanwj commented:
Tested ACK dfdf1d5

>  I think this is ready.

I agree. 
Are you going to do a final squash before merge ("Address some nits" and "Address more nits" seem like squashme's)?

harding commented:
How does this change affect UTXO set bloat attacks?

In the current model where UTXO entries are indexed by txid, I think the most efficient attack would be stuffing a block full of max-sized (10,000 byte?) scriptPubKeys.  That makes the worst-case UTXO byte-size increase per block slightly smaller than the maximum base block size.

But if UTXO entries are indexed by outpoint, it seems to me an attacker could create a single large transaction with many small outputs (minimum size 9 bytes?) which would each add a 33 to 36 byte compressed outpoint to the UTXO DB.  That makes the worst-case byte-size increase per block a multiple of the maximum base block size.

I see @sipa's comment in the OP about, "LevelDB internally uses an encoding that omits repeated prefix bytes in keys, and because of that, duplicating the txids is not very significant."  I was just wondering if the affect of this change on UTXO bloat attacks was considered (or was even worth considering)?

morcos commented:
This is a fantastic change.  I'm sorry it's taking me so long to work through it.  I can't promise I will finish today, but I have no objection to merge based on other's review and I'll keep working my way through either way.

@harding raises an interesting question, but I feel its more of an academic one, there is nothing preventing us from changing the implementation again in the future if that becomes an issue (or heaven forbid reverting)

gmaxwell commented:
@harding As you note that  sipa notes: the database deduplicates the IDs; this largely eliminates the concern your present.  The old structure also has quadratic access costs for transactions with many outputs which this change eliminates.

sipa commented:
@laanwj I'll squash soon.

@harding It's an interesting question what the worst cast UTXO bloat attack becomes after this change. I believe you're right that many small outputs will have a larger impact on disk size, though not as dramatically as you predict - most of it will be eaten by prefix compression.

sipa commented:
Pre-squash history:

* e66dbde6d14cb5f253b3bf8850a98f4fda2d9f49: Add SizeEstimate to CDBBatch
* f54580e7e4f225bb615204daef32f72ab8688418: error() in disconnect for disk corruption, not inconsistency
* e484652fc36ef7135cf08ad380ea7142b6cbadc0: Introduce CHashVerifier to hash read data
* 7e0032290669fae5f52c256856c53038511c7db4: Add specialization of SipHash for 256 + 32 bit data
* d342424301013ec47dc146a4beb49d5c9319d80a: Remove/ignore tx version in utxo and undo
* c3aa0c11947dfd82702df276d39bb7f748dd83a1: Report on-disk size in gettxoutsetinfo
* 7d991b55dbf0b0f6e21c0680ee3ebd09df09012f: Store/allow tx metadata in all undo records
* 422634e2f5ac1ff74cd358144cecbac63007adc4: Introduce Coin, a single unspent output
* 16a2789d33ab830787da8e716f4a65e14aecb159: Replace CTxInUndo with Coin
* 4d1246060c3587c1605f85735a50515929a8deff: Optimization: Coin&& to ApplyTxInUndo
* 7c1efc60f122ae18f501babdcae9047507267be6: Introduce new per-txout CCoinsViewCache functions
* b1216c967a9f12a18bcce9a9a2e39be30825c367: Switch from per-tx to per-txout CCoinsViewCache methods in some places
* 2cc9c5495c8c5f9b394017e163a21e4e3f9b2c84: Only pass things committed to by tx's witness hash to CScriptCheck
* f2d42ab5d45163aa2a6574884635ba3c2bb82984: Switch CScriptCheck to use Coin instead of CCoins
* 875b4123134c02a61486060996d3573116be776b: Switch tests from ModifyCoins to AddCoin/SpendCoin
* 96b992072a8cda6509b4185118aadb3097452ee2: Remove ModifyCoins/ModifyNewCoins
* f1b70c320f31d4a0cced7928757297cbb3475cff: Replace CCoins-based CTxMemPool::pruneSpent with isSpent
* 105377227ce041b943d149c59458d48319e56730: Refactor GetUTXOStats in preparation for per-COutPoint iteration
* 0dcdbbd0b36298299d59f3960cd8b0be48dd3b80: Switch CCoinsView and chainstate db from per-txid to per-txout
* b3cf86e59245d3715ab4e32186e0e30931f63aa8: Extend coins_tests
* d475d03c4ca1dedaba311add2e2c38279c79c6a0: Remove unused CCoins methods
* 363843fc4c0f1eadef589be20af34738e67ba835: Pack Coin more tightly
* c5faf390dfee5c969b20e8bee305542ef74fe8cb: Reduce reserved memory space for flushing
* 72eed2650fa9f3d31205712a965b9f99a5c8ee99: Upgrade from per-tx database to per-txout
* 7acbc2b7c688729c7c50802182300b7859f77349: [MOVEONLY] Move old CCoins class to txdb.cpp
* 854dad479c53960ee7fbae2bb7f544c10473b023: Merge CCoinsViewCache's GetOutputFor and AccessCoin
* ea0a7b9224d20962aa6bb78e5bd678b8d5492832: Rename CCoinsCacheEntry::coins to coin
* 1927df1a2316004d5cac22b65a561b578b37ec47: Increase travis unit test timeout
* 0328074fc823e0e9cc8461140df68fcdf70d43fd: scripted-diff: various renames for per-utxo consistency
* 1bc281f36e8ec88c66dad651c405b381cb9a39a0: Address some nits
* dfdf1d5fef732d24602e693f917e41ed357d4937: Address more nits

harding commented:
@morcos @gmaxwell @sipa Thanks for your responses (and, with the others, your work on this change!).  I agree the ability to adapt the underlying implementation to prevailing conditions makes the question academic, as hopefully does the increasing ratio of fees-to-subsidy which increases the lost-income cost to miners for maximizing bloat.

Please consider my comment non-blocking, and thank you all again for your work.

sipa commented:
Squashed all fixes without rebasing.

morcos commented on src/txdb.h:
```diff
@@ -22,9 +22,7 @@ class uint256;
-static constexpr int MAX_BLOCK_COINSDB_USAGE = 200 * DB_PEAK_USAGE_FACTOR;
-//! Always periodic flush if less than this much space still available.
-static constexpr int MIN_BLOCK_COINSDB_USAGE = 50 * DB_PEAK_USAGE_FACTOR;
+static constexpr int MAX_BLOCK_COINSDB_USAGE = 10 * DB_PEAK_USAGE_FACTOR;
```
maybe this should be 30?
sdaftuar said he saw 17 in only a week of testing

sdaftuar replied:
17 was incorrect; I mistakenly reported the cache delta between blocks as the per-block cache delta.

morcos reviewed (commented):
posthumous utACK (very untested)

nice work.
i left my couple of nits anyway.
I can open a PR to address the output lookups in ATMP and for any of the others that are worth addressing.

Earlz commented on src/rest.cpp:
```diff
@@ -42,16 +42,19 @@ static const struct {
     {
-        READWRITE(nTxVer);
+        uint32_t nTxVerDummy = 0;
+        READWRITE(nTxVerDummy);
```
What is the point of this? Just to keep the database backwards compatible?
